### PR TITLE
Glb import

### PR DESCRIPTION
**Pull request by najadojo**
Added an option to import a`.glb` binary package file.

### CODE REVIEW

**Comment by najadojo** [comment]
#33

**Comment by emackey** [comment]
This is amazing, I wasn't expecting it so soon!  I will still likely take a day or two to look over the code and test this with a couple different models.  I'm going to try to get this merged by Wednesday because it will be a fun topic for the glTF weekly call.  Thanks much!

File: src/extension.ts [hunk 1]
```diff
@@ -291,25 +292,34 @@ export function activate(context: vscode.ExtensionContext) {
...
-        };
-        vscode.window.showSaveDialog(options).then(uri => {
-            if (uri !== undefined) {
-                try {
-                    GlbExport.save(gltf, editor.document.uri.fsPath, uri.fsPath);
-                    vscode.window.showInformationMessage('Glb exported as: ' + uri.fsPath);
-                } catch (ex) {
-                    vscode.window.showErrorMessage(ex.toString());
+        if (fs.existsSync(glbPath)) {
+            const options: vscode.SaveDialogOptions = {
```
Comment (emackey) [inline src/extension.ts]: Do we need this change?  I imagine not every user will want a GLB file saved to the default filename in the same folder as the glTF file.  It makes for inconsistent UX if the app sometimes asks for a save location and sometimes not, depending on the presence of files that may not be in the user's immediate field of view.  I'd rather have the "Save As..." dialog always appear, I hope that's not too much of an inconvenience for your team.

File: src/importProvider.ts [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    }
+    if (!fs.existsSync(sourceFilename)) {
+        return;
+    }
+
+    // Compose an target filename
+    let targetFilename = sourceFilename.replace('.glb', '.gltf');
+
+    if (fs.existsSync(targetFilename)) {
+        let result = await vscode.window.showWarningMessage('.glTF already present, are you sure you want to continue?', 'Yes', 'Select Path');
```
Comment (emackey) [inline src/importProvider.ts]: Likewise here, let's just always show the "Save As..." dialog.

Comment (emackey) [reply src/importProvider.ts]: If there's a team that has a strong need for this kind of auto-acceptance of the default output filename, then I could imagine maybe a configuration setting to enable that for that team.  But for the majority of normal users, I would want to reliably let them see the file dialog.

File: src/extension.ts (hunk 1 above)
Comment (najadojo) [reply src/extension.ts]: I've added a configuration option as you suggested below.

File: src/importProvider.ts [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        bufferDataList.push(bufPart);
+    }
+
+    // go through all the buffer views and break out buffers as separate files
+    if (gltf.bufferViews !== undefined) {
+        for (let bufferViewIndex = 0; bufferViewIndex < gltf.bufferViews.length; bufferViewIndex++) {
+            let imageBuf = findImageBuf(bufferViewIndex);
+            if (imageBuf !== undefined) {
+                writeImageBuf(imageBuf, bufferViewIndex, binBuffer);
+                continue;
```
Comment (emackey) [inline src/importProvider.ts]: It looks like the old `bufferView` reference is kept in the imported glTF file, but never referenced since the image was pulled out to an external file.

I tested this by importing [`Lantern.glb`](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/Lantern/glTF-Binary) from the sample model repo, and diffing the results with [`Lantern.gltf`](https://github.com/KhronosGroup/glTF-Sample-Models/blob/master/2.0/Lantern/glTF/Lantern.gltf) from the same repo.  The resulting file is almost identical, but there are four extra, un-referenced bufferView entries, with sizes that look like these images.  The `accessors` section shows all bufferView indicies have increased by 4, and in particular index `0` to `3` are never referenced.

This actually isn't a big deal, the model appears to still work fine.  But it would be nice to pull these out if it's possible.

File: src/extension.ts [hunk 4]
```diff
@@ -354,6 +363,47 @@ export function activate(context: vscode.ExtensionContext) {
...
+                    vscode.window.showErrorMessage(ex.toString());
+                }
+            }
+        });
+    }));
+
+    context.subscriptions.push(vscode.commands.registerCommand('gltf.importGlbFile', (fileUri) => {
+
+        if (typeof fileUri == 'undefined' || !(fileUri instanceof vscode.Uri)) {
+            if (vscode.window.activeTextEditor === undefined) {
```
Comment (emackey) [inline src/extension.ts]: I'm struggling with this one.  It looks like `importGlbFile` maybe is looking for a GLB file open in the active editor?  If so, it fails, as `fileUri` and `activeTextEditor` are both `undefined` on my machine when a GLB is the active editor window.  When a GLTF file is active, things get worse, as the code below this attempts to read in the plain JSON as a GLB, and fails to find the correct magic number.

Comment (emackey) [reply src/extension.ts]: Maybe we don't need `importGlbFile`, only `importGlbOpen`?

File: src/extension.ts [hunk 5]
```diff
@@ -291,25 +291,34 @@ export function activate(context: vscode.ExtensionContext) {
...
-            }
-        };
-        vscode.window.showSaveDialog(options).then(uri => {
-            if (uri !== undefined) {
-                try {
-                    GlbExport.save(gltf, editor.document.uri.fsPath, uri.fsPath);
-                    vscode.window.showInformationMessage('Glb exported as: ' + uri.fsPath);
-                } catch (ex) {
-                    vscode.window.showErrorMessage(ex.toString());
+        if (!vscode.workspace.getConfiguration('glTF').get('neverPromptForFilename')) {
```
Comment (emackey) [inline src/extension.ts]: Thanks for this!  Minor nitpick, would you mind renaming to something like `alwaysOverwriteDefaultFilename`?  Maybe that's a bit wordy but I'm looking for something that has a positive sense (where true means yes) and not a negative sense (where true means don't do it).  This request stems from some unrelated software I worked on that had an option called "dontSave" that would fail to save when set to true, and confused users for years.

File: src/extension.ts [hunk 6]
```diff
@@ -231,40 +232,39 @@ export function activate(context: vscode.ExtensionContext) {
...
-                                // File exists, but user says it's OK to overwrite.
-                                exportToFile(filename, pathFilename, pointer, dataUri);
-                            }
-                        });
-                    } else {
-                        // File does not yet exist, try saving to it.
-                        exportToFile(filename, pathFilename, pointer, dataUri);
+                };
+                options.filters[mimeType] = [extension.replace('.', '')];
+                vscode.window.showSaveDialog(options).then(uri => {
```
Comment (emackey) [inline src/extension.ts]: Nice, thanks!

File: src/importProvider.ts [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+
+    // Read the GLB data
+    const Binary = {
+        Magic: 0x46546C67
+    };
+
+    const sourceBuf = fs.readFileSync(sourceFilename);
+    const readMagic = sourceBuf.readUInt32LE(0);
+    if (readMagic !== Binary.Magic) {
+        return;
```
Comment (emackey) [inline src/importProvider.ts]: Here, and everywhere in this file where you've encountered an error, we should display an error to the user.  Since this is wrapped in try/catch, we can *almost* do this simply by throwing the error from here:

```
        throw new Error('Source file does not appear to be a GLB (glTF Binary) model.');
```

But, this function is `async`, so to catch this error, the call site must await.  I'll leave a separate comment on that.

File: src/extension.ts [hunk 8]
```diff
@@ -354,6 +363,47 @@ export function activate(context: vscode.ExtensionContext) {
...
+        if (typeof fileUri == 'undefined' || !(fileUri instanceof vscode.Uri)) {
+            if (vscode.window.activeTextEditor === undefined) {
+                vscode.commands.executeCommand('gltf.importGlbOpen');
+                return;
+            }
+            fileUri = vscode.window.activeTextEditor.document.uri;
+        }
+
+        try {
+            GlbImport.load(fileUri.fsPath);
```
Comment (emackey) [inline src/extension.ts]: Here, we're calling `GlbImport.load`, which is async, so we must await if we want to catch errors that have happened after awaiting things in this function.

```
            await GlbImport.load(fileUri.fsPath);
```

File: src/extension.ts [hunk 9]
```diff
@@ -354,6 +363,47 @@ export function activate(context: vscode.ExtensionContext) {
...
+                try {
+                    GlbImport.load(fileUri[0].fsPath);
+                } catch (ex) {
+                    vscode.window.showErrorMessage(ex.toString());
+                }
+            }
+        });
+    }));
+
+    context.subscriptions.push(vscode.commands.registerCommand('gltf.importGlbFile', (fileUri) => {
```
Comment (emackey) [inline src/extension.ts]: After following my advice below about awaiting GlbImport.load, you'll have to declare this one as async.  I'm pretty sure this is the correct syntax, or at least it worked for me:

```
context.subscriptions.push(vscode.commands.registerCommand('gltf.importGlbFile', async (fileUri) => {
```

File: src/extension.ts [hunk 10]
```diff
@@ -354,6 +363,47 @@ export function activate(context: vscode.ExtensionContext) {
...
+             filters: {
+                'Binary glTF': ['glb'],
+                'All files': ['*']
+            }
+        };
+
+        vscode.window.showOpenDialog(options).then(fileUri => {
+            if (fileUri && fileUri[0]) {
+                try {
+                    GlbImport.load(fileUri[0].fsPath);
```
Comment (emackey) [inline src/extension.ts]: This should `await` the `.load` as well, in order to catch all errors.  See other notes in this review.

File: src/extension.ts (hunk 4 above)
Comment (najadojo) [reply src/extension.ts]: This I've been testing with the explorer bar... I haven't got it to work with the opened document.

Comment (emackey) [reply src/extension.ts]: Doh, I forgot to test with explorer bar.  Can this code auto-detect when the command was invoked from explorer vs. when it was invoked some other way?  If nothing else, look to see if `fileUri` actually ends with `.glb`, otherwise an open dialog will be needed.  That way the two import commands could be combined into one.

**Comment by emackey** [comment]
Thanks @najadojo.  I pushed some changes here, let me know if you have any concerns.  I hope to get this published within a day or two, it seems very close now.  `Lantern.glb` in particular imports to something that diffs to almost exactly the same non-binary version, which is stunning.  👍

**Comment by najadojo** [comment]
No concerns here.

**Comment by emackey** [comment]
Thanks so much @najadojo.  This will get published to the marketplace this weekend.  Super excited to have this in!

**Comment by emackey** [comment]
Version 2.0.8 is published!  🎉 

@najadojo What's your next amazing contribution?? 😸
